Invariant typevar matching on callsite checks

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25136 Invariant typevar matching on callsite checks**

Previously we were calling unifyType to match typevars at callsites.
unifyType actually does merging (e.g. to handle control flow joins)
so its effect at callsites was bivariance, allowing typevar bindings
to widen as new concrete types were encountered in arguments.

Fixes issue #24856

Strip refinements when doing invariant matching on type vars.

Previous change (bivariance to invariance) makes type matching
sensitive to the addition of type refinements. Use unshapedType
to avoid considering refinements when doing matching.

Differential Revision: [D17078081](https://our.internmc.facebook.com/intern/diff/D17078081)

Commit:
Invariant typevar matching on callsite checks

Previously we were calling unifyType to match typevars at callsites.
unifyType actually does merging (e.g. to handle control flow joins)
so its effect at callsites was bivariance, allowing typevar bindings
to widen as new concrete types were encountered in arguments.

Fixes issue #24856

Strip refinements when doing invariant matching on type vars.

Previous change (bivariance to invariance) makes type matching
sensitive to the addition of type refinements. Use unshapedType
to avoid considering refinements when doing matching.

zdevito reviewed (approved):
Looks good, thanks for tracking this down.

Commit:
Update on "Invariant typevar matching on callsite checks"


Previously we were calling unifyType to match typevars at callsites.
unifyType actually does merging (e.g. to handle control flow joins)
so its effect at callsites was bivariance, allowing typevar bindings
to widen as new concrete types were encountered in arguments.

Fixes issue #24856

Strip refinements when doing invariant matching on type vars.

Previous change (bivariance to invariance) makes type matching
sensitive to the addition of type refinements. Use unshapedType
to avoid considering refinements when doing matching.

Commit:
Update on "Invariant typevar matching on callsite checks"


Previously we were calling unifyType to match typevars at callsites.
unifyType actually does merging (e.g. to handle control flow joins)
so its effect at callsites was bivariance, allowing typevar bindings
to widen as new concrete types were encountered in arguments.

Fixes issue #24856

Strip refinements when doing invariant matching on type vars.

Previous change (bivariance to invariance) makes type matching
sensitive to the addition of type refinements. Use unshapedType
to avoid considering refinements when doing matching.
